lib/src/rewrite.rs

```diff
@@ -830,3 +830,133 @@ pub fn move_commits(
         num_skipped_rebases,
     })
 }
+
+pub struct CommitToSquash {
+    pub commit: Commit,
+    pub selected_tree: MergedTree,
+    parent_tree: MergedTree,
+}
+
+impl CommitToSquash {
+    pub fn new(repo: &dyn Repo, commit: Commit, selected_tree: MergedTree) -> BackendResult<Self> {
+        let parent_tree = commit.parent_tree(repo)?;
+        Ok(Self {
+            commit,
+            selected_tree,
+            parent_tree,
+        })
+    }
+
+    // Returns true if the selection contains all changes in the commit.
+    pub fn is_full_selection(&self) -> bool {
+        &self.selected_tree.id() == self.commit.tree_id()
+    }
+
+    // Returns true if the selection matches the parent tree (contains no changes
+    // from the commit).
+    //
+    // Both `is_full_selection()` and `is_empty_selection()`
+    // can be true if the commit is itself empty.
+    pub fn is_empty_selection(&self) -> bool {
+        self.selected_tree.id() == self.parent_tree.id()
+    }
+}
+
+pub struct SquashOptions {
```

Comment:
nit: add `#[derive(Clone, Debug)]`?

Reply:
Not possible anymore with the callback.

Reply:
It's probably easier to pass a callback by function argument. `Box<dyn` tends to make things complicated.

Reply:
Done.